# Welcome to Day 4! 

## Extracting and storing sequence data, working with GenBank files

### Section 1: Extracting and storing information from sequence data

### Section 2: Parsing and manipulating GenBank (.gbk/.gbff) files

---

## Session summary


Day 4 will have two different components. We will go over over how to extract sequence information, like GC content, lengnth, number of methionines, or whether a motif is present, and then save it to a file. We will also extract only information from the headers and save that to a file. Afterwards, we move into uncharted territory by learning how to work with GenBank files. Although GenBank files may look intimdating, we apply what we learned from working with Fasta files and realize that working with them can be straightforward.

---


# Section 1: Extracting and storing information from sequence data

Sometimes we might want to loop through sequences in a file and obtain sequence information, such as the GC content or the length of the sequence. We have obtained sequence information like this in Day 3 in order to filter for sequences, but now we wil learn how to store that information instead.

Here are some Biopython functions that can be used to obtain sequence information.

In [1]:
# import Biopython modules we will be using
from Bio import SeqIO
from Bio.Seq import Seq

In [3]:
# Calculate sequence length
for record in SeqIO.parse('antibiotic_resistance_genes_short.fasta','fasta'):
    print(len(record.seq))

744
729
1143


In [5]:
# Calculate number of AUG 
for record in SeqIO.parse('antibiotic_resistance_genes_short.fasta','fasta'):
    print(record.seq.count('ATG'))

9
12
22


In [6]:
# calculate GC content manually
for record in SeqIO.parse('antibiotic_resistance_genes_short.fasta','fasta'):
    c_count = record.seq.count('C')
    g_count = record.seq.count('G')
    seq_length = len(record.seq)
    print( 100 * (g_count+c_count)/seq_length )

30.64516129032258
28.9437585733882
47.41907261592301


We can save these data to a list or to a new file, depending on what we want to do. We can use the record.id so that each data point extracted has a unique identifier.

In [8]:
# make empty list
store_data = []

for record in SeqIO.parse('antibiotic_resistance_genes_short.fasta','fasta'):
    # each data point contains the record.id and the length in a list
    store_data.append([record.id, len(record.seq)])
# view data
store_data

[['ErmB', 744], ['ErmA', 729], ['ErmE', 1143]]

This list of lists data storage approach can be then converted into a Numpy array or a Pandas dataframe.

--- 
### Exercise 1a

We want to build a dataframe that contains information on whether specific protein motifs are located in each sequence. We will assign a 1 if it is there, and a 0 if it is not.

Use `SeqIO.parse()` to parse the fasta file and `translate()` the `record.seq`. `print()` each `record` in the loop.

In [ ]:
for ___ in ___.___('mixed_args.fasta', 'fasta'):
    record.seq = record.___.___
    print(___)

### Exercise 1b

To start, make `store_data` an empty list. 

Now let's search each `record` for the first motif `'QRALGL'` in our sequences using `find()`. Assign a 1 if it is found and a 0 if it is not. 

`append()` `sequence_identifier` and `motif_1` to `store_data`. 

Afterwards, `print()` `store_data`

In [ ]:
store_data = ___ 

for ___ in ___.___('mixed_args.fasta', 'fasta'):
    record.seq = record.___.___

    sequence_identifier = record.id

    if record.___.find('QRALGL') > -1:
        motif_1 = ___
    else:
        motif_1 = ___

    store_data.append([___, ___])

print(store_data)

### Exercise 1c 

Now let's test for the presence of two additional motifs, using the same basic code structure as above. Use `find()` to search for the other two motifs. Afterwards, `append()` `sequence_identifier`, `motif_1`, `motif_2`, and `motif_3` to `store_data`.

In [ ]:
store_data = ___ 

for ___ in ___.___('mixed_args.fasta', 'fasta'):
    record.seq = record.___.____

    sequence_identifier = record.id

    if ___.___.find('QRALGL') > -1:
        motif_1 = ___
    else:
        motif_1 = ___

    if ___.___.find('GAKR') > -1:
        motif_2 = ___
    else:
        motif_2 = ___

    if ___.___.find('DGTA') > -1:
        motif_3 = ___
    else:
        motif_3 = ___


    store_data.append([___, ___, ___, ___])

store_data

---

Sometimes it is necessary to extract data sequence headers themselves. An example would be taxonomy data that tells you where which species the sequence belongs to. The base python function `split()` is very useful here. If you know how the headers are organized, you can find common characters that can split the header, which is a string, into items in a list. 

As an example:



In [19]:
record.description = 'ErmB AAF86219.1 [Enterococcus faecium]'
# Splits record.description into four items in a list
record.description.split(' ')

['ErmB', 'AAF86219.1', '[Enterococcus', 'faecium]']

In [20]:
record.description = 'ErmB AAF86219.1 [Enterococcus faecium]'
# Splits record.description into two items in a list
record.description.split('[')

['ErmB AAF86219.1 ', 'Enterococcus faecium]']

We can use `split()` to convert the header into a list that we can then extract specific information from.

In [21]:
record.description = 'ErmB AAF86219.1 [Enterococcus faecium]'
header_data = record.description.split(' ') # header_data is now a list
genus = header_data[2] # retrieving the third item in the header_data list
species = header_data[3] # retrieving the fourth item in the header_data list

print(genus)
print(species)


[Enterococcus
faecium]


We can extract sequence information along with header information in a single pass.

In [22]:
# make empty list
store_data = []

for record in SeqIO.parse('antibiotic_resistance_genes_short.fasta','fasta'):

    # get individual counts of each nucleotide
    c_count = record.seq.count('C')
    g_count = record.seq.count('G')
    a_count = record.seq.count('A')
    t_count = record.seq.count('T')
    # get the length of the sequence
    seq_length = len(record.seq)
    # calculate the gc content of the sequence
    gc_content = 100 * (g_count+c_count)/seq_length

    # split the header by whitespace and store the resulting list in header_data
    header_data = record.description.split(' ')
    # get the genus of the sequence by accessing header_data
    genus = header_data[2]
    # get the species of the sequence by accessing header_data
    species = header_data[3]
     
    # store all of the data in store_data 
    store_data.append([record.id, seq_length, c_count, g_count, a_count, t_count, gc_content, genus, species])

# view data
store_data

[['ErmB',
  744,
  120,
  108,
  297,
  219,
  30.64516129032258,
  '[Enterococcus',
  'faecium]'],
 ['ErmA',
  729,
  103,
  108,
  298,
  220,
  28.9437585733882,
  '[Staphylococcus',
  'aureus]'],
 ['ErmE',
  1143,
  214,
  328,
  401,
  200,
  47.41907261592301,
  '[Saccharopolyspora',
  'erythraea']]

---
### Exercise 2a

Among other things, we want to extract the genus and species from each sequences in `mixed_args.fasta` 

Before we try to extract data from the headers, we want to examine how the headers are organized to see where we can `split()` them.

Use `SeqIO.parse` to loop through each `record` in `mixed_args.fasta`.

`print()` each `record.description`

In [ ]:
for record in SeqIO.___('mixed_args.fasta', 'fasta'):
    print( ___ )

### Exercise 2b

The headers (`record.description`) are more complicated than the previous example. 

However, a good place to `split()` the `record.description` would be the `=` signs.

Store the resulting list in the variable `header_data`. 

`print()` `header_data` to see how `record.description` was split.




In [29]:
for record in SeqIO.parse('mixed_args.fasta', 'fasta'):
    header_data = record.description.___('___')
    print(header_data)
    print('\n')

['ERMA_STAAR Q6GKQ0 rRNA adenine N-6-methyltransferase OS', 'Staphylococcus aureus (strain MRSA252) OX', '282458 GN', 'ermA1 PE', '3 SV', '1']
['ERMA_STAAM P0A0H1 rRNA adenine N-6-methyltransferase OS', 'Staphylococcus aureus (strain Mu50 / ATCC 700699) OX', '158878 GN', 'ermA1 PE', '3 SV', '1']
['ERMA_STAAU P0A0H3 rRNA adenine N-6-methyltransferase OS', 'Staphylococcus aureus OX', '1280 GN', 'ermA PE', '3 SV', '1']
['BLAC_MYCTO P9WKD2 Beta-lactamase OS', 'Mycobacterium tuberculosis (strain CDC 1551 / Oshkosh) OX', '83331 GN', 'blaC PE', '3 SV', '1']
['BLAC_MYCBO P0A5I7 Beta-lactamase OS', 'Mycobacterium bovis (strain ATCC BAA-935 / AF2122/97) OX', '233413 GN', 'blaC PE', '3 SV', '1']
['A0A133QE03_STASI Erythromycin resistance protein OS', 'Staphylococcus simulans OX', '1286 GN', 'HMPREF3215_02510 PE', '3 SV', '1']
['AMPC_CITFR P05193 Beta-lactamase OS', 'Citrobacter freundii OX', '546 GN', 'ampC PE', '1 SV', '1']
['D4REG1_ENTFC D4REG1 Macrolide-lincosamide-streptogramin B resistance p

### Exercise 2c

Using the first `split()` `record.description` as a reference, here is the structure of the splitted list.

`['ERMA_STAAR Q6GKQ0 rRNA adenine N-6-methyltransferase OS', 'Staphylococcus aureus (strain MRSA252) OX', '282458 GN', 'ermA1 PE', '3 SV', '1']`

We can see that the taxonomic data is the second item in the list (or first if we use Python indexing). 

For each split `record.description`, `append()` the item with the `taxonomic_info` to the list `taxonomy_store`.


In [ ]:
taxonomy_store = ___

for record in SeqIO.parse('mixed_args.fasta', 'fasta'):
    header_data = record.description.___('___')
    taxonomic_info = header_data[___]
    ___.append(___)

taxonomy_store

---

We can use `SeqIO` for more than just fasta files. For example, we can examine the reads in `.fastq` file using almost the same code as for a fasta file.

In [31]:
for record in SeqIO.parse('short_reads.fastq', 'fastq'):
    print(record)

ID: M05959:27:000000000-D8Y5P:1:1102:15343:1546
Name: M05959:27:000000000-D8Y5P:1:1102:15343:1546
Description: M05959:27:000000000-D8Y5P:1:1102:15343:1546 1:N:0:CGCTGT
Number of features: 0
Per letter annotation for: phred_quality
Seq('TTTTCTTCTTTCTTTCCTTTCTCTTTTTCCCCCATCTCAACGTATTATCCTTTT...TCA', SingleLetterAlphabet())
ID: M05959:27:000000000-D8Y5P:1:1102:15343:1546
Name: M05959:27:000000000-D8Y5P:1:1102:15343:1546
Description: M05959:27:000000000-D8Y5P:1:1102:15343:1546 2:N:0:CGCTGT
Number of features: 0
Per letter annotation for: phred_quality
Seq('CTTTGTTTTTTTTTTAGTGGGGTAGATTGGGAAAGGCAAGTCTGTAATATTGAA...AAA', SingleLetterAlphabet())
ID: M05959:27:000000000-D8Y5P:1:1102:16544:1615
Name: M05959:27:000000000-D8Y5P:1:1102:16544:1615
Description: M05959:27:000000000-D8Y5P:1:1102:16544:1615 1:N:0:CGATGT
Number of features: 0
Per letter annotation for: phred_quality
Seq('TCTGTAGCACCGCCTACATACCTCGCTCTGCTAATCCTGTTACCAGTGGCTGCT...TTT', SingleLetterAlphabet())
ID: M05959:27:000000000-D8Y5P:1

---
### Exercise 3a



`print()` the `read_length` of each read in `short_reads.fast` using `len()`.

In [ ]:
for record in SeqIO.parse('short_reads.fastq', '___'):
    read_length = ___(record.___)
    print(read_length)

### Exercise 3b
count the **total** number of reads in `short_reads.fastq`. `print()` the total number of reads.

In [ ]:
total_count = 0

for record in SeqIO.parse('short_reads.fastq', '___'):
    total_count = total_count + ___

print('total reads:', ___)

### Exercise 3c

`print()` the `record.id` and `read_length` of all reads that have fewer than or equal to 270 bases.

In [170]:
for record in SeqIO.parse('short_reads.fastq', '___'):
    read_length = ___(record.___)
    if read_length <= ___:
        print('ID:', ___.id)
        print('length:', ___)
        print('\n')

262
269
272
300


---



## Interlude

Before we get to the next section, we need to review dictionaries.

Dictionaries are similar to lists in that they are a way to hold data. A dictionary is different because it is composed of a 'key' and 'value' pairs. For example


In [124]:
my_dict = { 'animal': ['dog','cat'],
            'person': ['taylor','simone'],
            'vehicle': ['car','plane']}

my_dict

{'animal': ['dog', 'cat'],
 'person': ['taylor', 'simone'],
 'vehicle': ['car', 'plane']}

You can access the **elements** in a key by using they keyword

In [128]:
# access all elements in key 'animal'
my_dict['animal']

['dog', 'cat']

In [ ]:
# access the first element of key 'animal'
my_dict['animal'][0]

# Section 2: Parsing and manipulating GenBank (.gbk/.gbff) files

Until now we have been exclusively working with sequences in `fasta` format. As we saw from our most recent example, however, `SeqIO` works with [`fastq` files](https://www.researchgate.net/profile/Morteza-Hosseini-6/publication/309134977/figure/fig2/AS:417452136648711@1476539753452/A-sample-of-the-FASTQ-file_W640.jpg). As we're finding out, `SeqIO` also works with other file types, including `GenBank` files. 

Compared to [`fasta` files](https://www.researchgate.net/profile/Morteza-Hosseini-6/publication/309134977/figure/fig1/AS:417452136648705@1476539753111/A-sample-of-the-Multi-FASTA-file_W640.jpg), [`GenBank` files](https://archaea.i2bc.paris-saclay.fr/baget/Images/hup.png) contain more information.

We will go over how to open and explore the data found in a `GenBank` file as well as how to extract features of interest and save them to a new file.

We will start by opening a `.gbk` file that only has one gene in it called fabV. We can see that we can access the same types of information as we would a `fasta` file.

In [57]:
for record in SeqIO.parse('fabv_small.gbk', 'genbank'):
    # record id
    print(record.id)
    print('\n')
    print(record.description)
    print('\n')
    print(record.seq)

WP_003102929.1


MULTISPECIES: trans-2-enoyl-CoA reductase family protein [Pseudomonas]


MIIKPRVRGFICVTTHPAGCEANVKQQIDYVEAKGPVVNGPKKVLVIGSSTGYGLAARITAAFGSGADTLGVFFERPGSESKPGTAGWYNSAAFEKFAHEKGLYARSINGDAFSDEVKRLTIETIKRDLGKVDLVVYSLAAPRRTHPKSGEVFSSTLKPIGKSVSFRGLDTDKEVIKDVVLEAASDQEVADTVAVMGGEDWQMWIDALLEADVLADGAKTTAFTYLGEKITHDIYWNGSIGAAKKDLDQKVLGIRDKLAPLGGDARVSVLKAVVTQASSAIPMMPLYLSLLFKVMKEQGTHEGCIEQVDGLYRESLYGAEPRLDEEGRLRADYKELQPEVQSRVEELWDKVTNENLYELTDFAGYKSEFLNLFGFEVAGVDYEQDVNPDVQIANLIQA


But there additional things we can access, since GenBank files are more information-rich.

The most important new thing we can access is called `features`. You would call upon it by inputting `record.features`

The data stored in record.features `record.features` is actually a list of dictionaries. We can loop through `record.features` to examine its contents:

In [68]:
for feature in record.features:
    print(feature)
    print('\n')

type: source
location: [0:398]
qualifiers:
    Key: db_xref, Value: ['taxon:286']
    Key: organism, Value: ['Pseudomonas']



type: Protein
location: [0:398]
qualifiers:
    Key: calculated_mol_wt, Value: ['43397']
    Key: product, Value: ['trans-2-enoyl-CoA reductase family protein']



type: Region
location: [0:396]
qualifiers:
    Key: db_xref, Value: ['CDD:237460']
    Key: note, Value: ['enoyl-[acyl-carrier-protein] reductase FabV']
    Key: region_name, Value: ['PRK13656']





From the above, we can see that each loop returns three basic things: type, location, and qualifiers. 

In [73]:
# what is the specific feature for?
for feature in record.features:
    print(feature.type)

source
Protein
Region


In [74]:
# where in the genome/sequence is the feature of interest located?
for feature in record.features:
    print(feature.location)

[0:398]
[0:398]
[0:396]


In [72]:
# the special data associated with the feature
for feature in record.features:
    print(feature.qualifiers)

OrderedDict([('organism', ['Pseudomonas']), ('db_xref', ['taxon:286'])])
OrderedDict([('product', ['trans-2-enoyl-CoA reductase family protein']), ('calculated_mol_wt', ['43397'])])
OrderedDict([('region_name', ['PRK13656']), ('note', ['enoyl-[acyl-carrier-protein] reductase FabV']), ('db_xref', ['CDD:237460'])])


We can access specific information found in `record.features`. **But** we have to make sure we specify the correct `feature.type` first.

In [75]:
# the organism that the sequence belongs to is stored under feature type called 'source'
for feature in record.features:
    if feature.type == 'source':
        print(feature.qualifiers['organism'])

['Pseudomonas']


In [77]:
# the molecular weight is stored under the feature type 'Protein'
for feature in record.features:
    if feature.type == 'Protein':
        print(feature.qualifiers['calculated_mol_wt'])

['43397']


---
### Exercise 4a

`print()` each `feature` found in `record.features` in `'fabv_small.gbk'`.

In [80]:
for record in SeqIO.parse('fabv_small.gbk', '___'):
    for feature in record.___:
        print(feature)

type: source
location: [0:398]
qualifiers:
    Key: db_xref, Value: ['taxon:286']
    Key: organism, Value: ['Pseudomonas']

type: Protein
location: [0:398]
qualifiers:
    Key: calculated_mol_wt, Value: ['43397']
    Key: product, Value: ['trans-2-enoyl-CoA reductase family protein']

type: Region
location: [0:396]
qualifiers:
    Key: db_xref, Value: ['CDD:237460']
    Key: note, Value: ['enoyl-[acyl-carrier-protein] reductase FabV']
    Key: region_name, Value: ['PRK13656']



### Exercise 4b

`print()` all the `feature.qualifiers` found in `feature.type` `'Region'`

In [ ]:
for record in SeqIO.parse('fabv_small.gbk', '___'):
    for feature in record.___:
        if feature.___ == 'Region':
            print(feature.___)

### Exercise 4c

`print()` the database cross reference (`'db_xref'`) identifier stored in `feature.qualifiers` that belongs to `feature.type` `'Region'`

In [ ]:
for record in SeqIO.parse('fabv_small.gbk', '___'):
    for feature in record.___:
        if feature.___ == 'Region':
            print(feature.___['___'])

---

After working with fasta files for so long, it might seem that GenBank files are un-necessarily complex, even if its structure can hold more information. And it is true that perhaps for just one gene/protein, we wouldn't want to access the sequence in GenBank format. 

Where the GenBank format truly shines is for storing assembled whole genomes that are annotated with predicted coding regions, tRNA, rRNA, and other features.


In [86]:
# record-level data will give you the name of the genome and also the nucleotide sequence for the whole assembly. Each contig is a new record.
for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    print(record.description)
    print(len(record.seq))
    print('\n') # for clarity

Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome
6823040


Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 plasmid pPA52Ts32, complete sequence
1584




From the above output, we can see that we have a really well-assembled genome. It only has two contigs. The first is the complete chromosomal genome and the second is a circular plasmid. When we examine the lengths, we notice there is a big discrepancy in the number of basepairs each has.


We can easily extract some useful metadata about the genomes features. We start by input `record.annotations`. This is will make a wall of text but don't be afraid. It is simply a big dictionary that has a couple of dictionaries within it.

In [111]:
for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    rec_annot = record.annotations

rec_annot

{'molecule_type': 'DNA',
 'topology': 'circular',
 'data_file_division': 'CON',
 'date': '12-AUG-2020',
 'accessions': ['NZ_CP059062'],
 'sequence_version': 1,
 'keywords': ['RefSeq'],
 'source': 'Pseudomonas aeruginosa',
 'organism': 'Pseudomonas aeruginosa',
 'taxonomy': ['Bacteria',
  'Proteobacteria',
  'Gammaproteobacteria',
  'Pseudomonadales',
  'Pseudomonadaceae',
  'Pseudomonas'],
 'references': [Reference(title='Direct Submission', ...)],
 'comment': 'REFSEQ INFORMATION: The reference sequence is identical to\nCP059062.1.\nThe annotation was added by the NCBI Prokaryotic Genome Annotation\nPipeline (PGAP). Information about PGAP can be found here:\nhttps://www.ncbi.nlm.nih.gov/genome/annotation_prok/\nCOMPLETENESS: full length.',
 'structured_comment': OrderedDict([('Genome-Assembly-Data',
               OrderedDict([('Assembly Method',
                             'CLC Genomics Workbench v. 20; SPAdes v. 3.13.0'),
                            ('Genome Representation', 'Full')

The real meat containing the summary of the overall genomic features is found under `record.annotations['structured_comment']['Genome-Annotation-Data']`

In [113]:
for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    rec_annot_metadata = record.annotations['structured_comment']['Genome-Annotation-Data']

rec_annot_metadata

OrderedDict([('Annotation Provider', 'NCBI RefSeq'),
             ('Annotation Date', '08/11/2020 21:20:25'),
             ('Annotation Pipeline',
              'NCBI Prokaryotic Genome Annotation Pipeline (PGAP)'),
             ('Annotation Method',
              'Best-placed reference protein set; GeneMarkS-2+'),
             ('Annotation Software revision', '4.12'),
             ('Features Annotated',
              'Gene; CDS; rRNA; tRNA; ncRNA; repeat_region'),
             ('Genes (total)', '6,302'),
             ('CDSs (total)', '6,222'),
             ('Genes (coding)', '6,136'),
             ('CDSs (with protein)', '6,136'),
             ('Genes (RNA)', '80'),
             ('rRNAs', '4, 4, 4 (5S, 16S, 23S)'),
             ('complete rRNAs', '4, 4, 4 (5S, 16S, 23S)'),
             ('tRNAs', '64'),
             ('ncRNAs', '4'),
             ('Pseudo Genes (total)', '86'),
             ('CDSs (without protein)', '86'),
             ('Pseudo Genes (ambiguous residues)', '0 of 86'),


We can extract specific items from this dictionary using keywords.

In [106]:
# number of coding sequences
for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    n_coding_sequences = record.annotations['structured_comment']['Genome-Annotation-Data']['CDSs (total)']

n_coding_sequences

'6,222'

In [107]:
# number of CRISPR arrays
for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    n_crispr_arrays = record.annotations['structured_comment']['Genome-Annotation-Data']['CRISPR Arrays']

n_crispr_arrays

'3'

We can also access other parts of `record.annotations`


In [116]:
# full taxonomic breakdown
for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    genome_taxonomy = record.annotations['taxonomy']

genome_taxonomy

['Bacteria',
 'Proteobacteria',
 'Gammaproteobacteria',
 'Pseudomonadales',
 'Pseudomonadaceae',
 'Pseudomonas']

----

### Exercise 5a:

Retrieve the list of database cross-reference identifiers stored in `record.dbxrefs`. 


In [142]:
for record in SeqIO.___('salmonella_genome.___','genbank'):
    genome_assembly_ids = ___.dbxrefs

genome_assembly_ids

['BioProject:PRJNA224116',
 'BioSample:SAMN15501998',
 'Assembly:GCF_014109785.1']

### Exercise 5b:

from `genome_assembly_ids`, select the `'Assembly'` ID. This is the whole genome identifer for the genome in NCBI

In [143]:
___[2]

'Assembly:GCF_014109785.1'

### Exercise 5c:

Retrieve the number of pseudogenes stored in `record.annotations['structured_comment']['Genome-Annotation-Data']` using the keyword `'Pseudo Genes (total)'`


In [ ]:
for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    n_pseudo_genes = record.___['structured_comment']['Genome-Annotation-Data']['___']

n_pseudo_genes

### Exercise 5d:

Use `len()` to retrieve the length of the **whole** genome. 

Remember that each `record` object provides information for each contig. Our genome has **two** contigs, so we will need to add up the lengths of `record.seq`

In [152]:
total_genome_length = ___

for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    contig_length = len(___.___)
    total_genome_length = total_genome_length + ___

total_genome_length

6824624

### Exercise 5e:

Being able to to quickly go through a genome and extract relevant metadata is a common bioinformatic task. 

We will put everything we learned in the previous steps of exercise 5 together into a single step. Look at the comments for hints

Extract the following metadata: 

1. NCBI whole-genome id 
2. number of genes
3. number of pseudogenes
4. the length of the genome.

`append()` the extracted metadata to store_data in that order.

In [ ]:
store_data = ___

total_genome_length = ___

for record in SeqIO.parse('salmonella_genome.gbk','genbank'):

    # ncbi whole-genome id stored in record
    genome_assembly_ids = ___.dbxrefs
    ncbi_whole_genome_id = ___[2]

    # number of genes stored in record.annotations under keyword 'Genes (coding)'
    n_genes = record.___['structured_comment']['Genome-Annotation-Data']['Genes (coding)']

    # number of pseudogenes stored in record.annotations under keyword 'Pseudo Genes (total)'
    n_pseudo_genes = record.___['structured_comment']['Genome-Annotation-Data']['___']

    # length of the genome from getting the len() of record.seq
    contig_length = len(record.___)
    total_genome_length = total_genome_length + ___

store_data.append(___) #whole genome id
store_data.append(___) #number of genes
store_data.append(___) #number of pseudogenes
store_data.append(___) #length of whole genome

store_data

---

One last thing to go over are the record.features in a large GenBank file. 

Let's say we wanted to manually count the number of coding sequences 



In [89]:
# record-level data will give you the name of the genome and also the nucleotide sequence for the whole assembly. Each contig is a new record.
cds_count = 0

for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    for feature in record.features:
        if feature.type == 'CDS':
            cds_count = cds_count + 1

cds_count

6222

Or, let's say we want to make a list of all the gene products descriptions in our genome for further analysis

In [160]:
product_list = []

for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    for feature in record.features:
        if feature.type == 'CDS':
            product_list.append(feature.qualifiers['product'][0])

product_list

ux transporter periplasmic adaptor subunit',
 'ribosome-associated ATPase/putative transporter RbbA',
 'ABC transporter permease',
 'EVE domain-containing protein',
 '5-formyltetrahydrofolate cyclo-ligase',
 'cell division protein ZapA',
 'TIGR02449 family protein',
 'YecA family protein',
 'Xaa-Pro aminopeptidase',
 '2-octaprenyl-6-methoxyphenyl hydroxylase',
 'YiiD C-terminal domain-containing protein',
 '2-octaprenyl-3-methyl-6-methoxy-1,4-benzoquinol hydroxylase',
 'hypothetical protein',
 'MFS transporter',
 'LysR family transcriptional regulator',
 'extracellular solute-binding protein',
 'iron ABC transporter permease',
 'glycine cleavage system aminomethyltransferase GcvT',
 'glycine cleavage system protein GcvH',
 'glycine dehydrogenase (aminomethyl-transferring)',
 'DUF2388 domain-containing protein',
 'NfeD family protein',
 'type II/IV secretion system protein',
 'inorganic triphosphatase',
 'TIGR00153 family protein',
 'inorganic phosphate transporter',
 'acetylornithine d

There are a lot of products here. What if we wanted to find only the ones that are involved in conjugal transfer? Also, are these genes located in the chromosomal or plasmid DNA, or both?

In [165]:
conjugal_transfer_list = []

for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    for feature in record.features:
        if feature.type == 'CDS':
            if feature.qualifiers['product'][0].lower().find('conjugal') > -1:
                conjugal_transfer_list.append([record.description,feature.qualifiers['product'][0]])

conjugal_transfer_list

[['Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome',
  'conjugal transfer protein TraG'],
 ['Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome',
  'conjugal transfer protein TrbE'],
 ['Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome',
  'conjugal transfer protein TrbF'],
 ['Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome',
  'conjugal transfer protein TraG N-terminal domain-containing protein'],
 ['Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome',
  'TIGR03751 family conjugal transfer lipoprotein'],
 ['Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome',
  'TIGR03750 family conjugal transfer protein'],
 ['Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome',
  'conjugal transfer protein TraX'],
 ['Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome',
  'TIGR03750 family conjugal transfer protein'],
 ['Pseudomonas aeruginosa strain GIMC5034:PA52Ts32 chromosome',
  'TIGR03751 family conjugal transfer lipoprotein']

Surprisingly, conjujgal transfer genes were only found on the chromosome. In the next exercise, let's get all the gene products found in the plasmid.

---

### Exercise 6a:

As has been stated before, in a whole-genome GenBank file, each record is a contig in the genome. In our genome .gbk, there are only two contigs, chromosomal and plasmid contigs. 

`print()` the `record.id` of each contig.



In [166]:
for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    print(___)

NZ_CP059063.1
NZ_CP059062.1


### Exercise 6b:

The second contig (the plasmid contig) has the record.id `NZ_CP059062.1`



Let's set our loop so that when the `record.id` is `NZ_CP059062.1`, a variable called `product_description` is created from each `feature.qualifiers['product'][0]` when `feature.type` is `CDS`.  

`append()` the `product_description` to the list `plasmid_products`

In [167]:
plasmid_products = ___

for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    if record.id == '___':
        for feature in record.features:
            if feature.type == 'CDS':
                product_description = feature.qualifiers['___'][0]
                ___.append(product_description)

plasmid_products

['hypothetical protein',
 'replication initiation factor domain-containing protein']

### Exercise 6c:

Do the same as in exercise 6b, but also extract the `ncbi_accession`, which is the gene ID, that is stored in `feature.qualifiers' under the keyword 'locus_tag'.

`append()` both the `product_description` and `ncbi_accession` to `plasmid_products`

In [168]:
plasmid_products = ___

for record in SeqIO.parse('salmonella_genome.gbk','genbank'):
    if record.id == '___':
        for feature in record.features:
            if feature.type == 'CDS':
                product_description = feature.qualifiers['___'][0]
                ncbi_accession = feature.___['locus_tag'][0]
                
                plasmid_products.append([___, ___])

plasmid_products

[['PA52Ts32_RS00005', 'hypothetical protein'],
 ['PA52Ts32_RS00010',
  'replication initiation factor domain-containing protein']]

---

That's it for Day 4! 